enhancements/authentication/oauth-resource-storage.md

```diff
@@ -0,0 +1,253 @@
+    // in order to authorize with the API server. It is concatenated with the name of the 
+    // OAuthAuthorizeToken object.
+    // For old tokens, this field might be empty.
+    // +sensitive
+    // +optional
+    SecretToken string `json:"secretToken,omitempty" protobuf:"bytes,11,opt,name=secretToken"`
+}
+```
+
+### User Stories [optional]
+
+#### Story 1
+
+An attacker who got access to an etcd backup I do not get access to oauth tokens.
+
+#### Story 2
+
+As a user I can use the API via oc or console to log out other sessions. 
+
+#### Story 3
+
+(optional) As a user I can list my other oauth sessions, including information like client name and expiration.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The oauth authorizer in kube-apiserver in v4.6 will split the provided bearer token into 
+the first 43 bytes (256 bits, compare https://github.com/openshift/oauth-server/blob/3539e4065084dc0b5f35e43d7ea49631df99760d/pkg/server/crypto/random.go#L31-L35)
+and remaining bytes. It then looks up the `OAuthAccessToken` via the 
+former as a name (as we do today) and in addition compares the `SecretToken` field with
+the latter (possible zero length).   
+
+Old tokens will have an empty `SecretToken` field, and clients using that token will only
+pass 43 bytes bearer token, leading into the zero length case.
+
+Old kube-apiserver (on downgrade) will try to lookup the token object via the whole token
+and fail. I.e. on downgrade clients have to reauthenticate.
+
+#### Logout and Read Access for Users
+
+With the `SecretToken`, `AuthorizeToken` and `RefreshToken` fields hidden for non-admin users,
+we can allow (via RBAC) normal users to GET, LIST, WATCH, DELETE their own tokens. This can 
+be implemented by injecting a label selector in the List and Watch implementation of the `OAuthAccessToken` 
+registry, similarly as done when proxying project requests to namespaces (where the user
+can list only his own projects; compare https://github.com/openshift/openshift-apiserver/blob/4c982be0e64177b727200e416aa6375f4f17ffc2/pkg/project/apiserver/registry/project/proxy/proxy.go#L83).
+
+For this list by label to work, we have to label new `OAuthAccessToken` with their owner, i.e.
+
+```
+oauth.openshift.io/username: <user-name-matching-the-context-user>
+```
+
+Old tokens won't have this label and won't be visible. As tokens expire after a day, this is acceptable.
+
+### Risks and Mitigations
+
+- When we adds an RBAC role and binding to read OAuthAccessTokens on v4.6 and then downgrade the
+  cluster to v4.5, this RBAC role and binding will allow to read **all** tokens. We can mitigate this
+  by adding the upper label based registry logic to a v4.5 z-release. 
+  
+### Test Plan
+
+- Old tokens with new kube-apiserver after v4.5->v4.6 upgrade have to keep working.
+- New tokens with an old kube-apiserver after a v4.6->v4.5 downgrade will refuse authentication.
+
+### Upgrade / Downgrade Strategy
```

Comment:
Can we have a mix of old and new Kube instances during an upgrade?
If yes, then it might fail when a token in the new format will arrive at the old instance.

Reply:
Good point. What does this mean? We have to delay creation of new token until we know that we have upgraded completely, right?

Reply:
oauth-server creates the tokens. Do we have an order of upgrades which helps here or is it completely arbitrary? cc @stlaz @deads2k

Reply:
> oauth-server creates the tokens. Do we have an order of upgrades which helps here or is it completely arbitrary? cc @stlaz @deads2k

There is no order of upgrades.  You can either put in toleration logic in 4.5 and deliver the full change in 4.6 or you can create a flag that is twiddled based on state of another component.  I suggest the two release update cadence.

Reply:
> There is no order of upgrades. You can either put in toleration logic in 4.5 and deliver the full change in 4.6 or you can create a flag that is twiddled based on state of another component. I suggest the two release update cadence.

Or you can retcon the new token validation into the older level of kube-apiserver.  I'd be supportive of that.

Reply:
We will add sha256 verification capabilities to v4.n-1 z-stream.